framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,280 @@
+     */
+    public boolean returnsThis(final MethodInvocationTree tree) {
+        ReturnsReceiverAnnotatedTypeFactory rrATF =
+                getTypeFactoryOfSubchecker(ReturnsReceiverChecker.class);
+        ExecutableElement methodEle = TreeUtils.elementFromUse(tree);
+        AnnotatedTypeMirror methodAtm = rrATF.getAnnotatedType(methodEle);
+        AnnotatedTypeMirror rrType =
+                ((AnnotatedTypeMirror.AnnotatedExecutableType) methodAtm).getReturnType();
+        return rrType != null && rrType.hasAnnotation(This.class);
+    }
+
+    /**
+     * Is the given annotation an accumulator annotation?
+     *
+     * @param anm an annotation mirror
+     * @return whether the annotation mirror is an instance of this factory's accumulator annotation
+     */
+    public boolean isAccumulatorAnnotation(AnnotationMirror anm) {
+        return AnnotationUtils.areSameByClass(anm, ACC);
+    }
+
+    @Override
+    protected TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(
+                super.createTreeAnnotator(), new AccumulationTreeAnnotator(this));
+    }
+
+    /** Handles fluent APIs using the Returns Receiver Checker. */
+    protected class AccumulationTreeAnnotator extends TreeAnnotator {
+
+        /**
+         * Mandatory constructor.
+         *
+         * @param factory the type factory
+         */
+        public AccumulationTreeAnnotator(AccumulationAnnotatedTypeFactory factory) {
+            super(factory);
+        }
+
+        @Override
+        public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror type) {
+            // Check to see if the ReturnsReceiver Checker has a @This annotation
+            // on the return type of the method.
+            if (returnsThis(tree)) {
+
+                // Fetch the current type of the receiver, or top if none exists.
+                ExpressionTree receiverTree = TreeUtils.getReceiverTree(tree.getMethodSelect());
+                AnnotatedTypeMirror receiverType;
+                AnnotationMirror receiverAnno;
+
+                if (receiverTree != null
+                        && (receiverType = getAnnotatedType(receiverTree)) != null) {
+                    receiverAnno = receiverType.getAnnotationInHierarchy(TOP);
+                } else {
+                    receiverAnno = TOP;
+                }
+
+                type.replaceAnnotation(receiverAnno);
+            }
+            return super.visitMethodInvocation(tree, type);
+        }
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(MultiGraphFactory factory) {
+        return new AccumulationQualifierHierarchy(factory);
+    }
+
+    /**
+     * All accumulation analyses share a similar type hierarchy. This hierarchy implements the
+     * subtyping, LUB, and GLB for that hierarchy. The lattice looks like:
+     *
+     * <p>top / \ acc(x) acc(y) ... \ / acc(x,y) ... | bottom
```

Comment:
FWIW I can't really read this lattice.  Is this a standard notation?

Reply:
The formatter messed this up. Sorry about that.